### ISSUE TITLE
Does not work with newest takeout :-(

### ISSUE THREAD

**Issue by wunderbaum** [kpym/hist2geo#2]
Hello,
I just downloaded my Takeout and tried to convert it to kml. Unfortunately, it did not work.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x18 pc=0xb630a3]

goroutine 1 [running]:
bufio.(*Reader).Read(0xc0001980c0, {0xc179b82000, 0x1ff, 0x225fdaff248?})
        /opt/hostedtoolcache/go/1.23.1/x64/src/bufio/bufio.go:241 +0x183
github.com/goccy/go-json/internal/decoder.(*Stream).read(0xc000198120)
        /home/runner/go/pkg/mod/github.com/goccy/go-json@v0.10.2/internal/decoder/stream.go:217 +0x158
github.com/goccy/go-json/internal/decoder.(*Stream).Token(0xc000198120)
        /home/runner/go/pkg/mod/github.com/goccy/go-json@v0.10.2/internal/decoder/stream.go:172 +0x5b
github.com/goccy/go-json.(*Decoder).Token(...)
        /home/runner/go/pkg/mod/github.com/goccy/go-json@v0.10.2/decode.go:245
main.Read({0xd82cc8, 0xc0001980c0})
        /home/runner/work/gotoextr/gotoextr/main.go:96 +0x14c
main.main()
        /home/runner/work/gotoextr/gotoextr/main.go:260 +0xa88
```

So I extracted the file _Timeline Edits.json_ ( about 33MB size) and renamed it to Timeline.json (because of the space) as I have not found a records.json.  There I just get an EOL output.

Any hope to get my tracks since 2014 converted?
Thank you!

**Comment by kpym** [comment]
@wunderbaum thank you for reporting this issue and for providing detailed information about the crash.

### Analysis of the Problem
Since **2024**, Google no longer saves location history directly on their servers. Instead, this data is stored locally on users' phones. Consequently, the `.zip` file downloaded via Google Takeout may be nearly empty.

Additionally, the current `.json` file structure exported from phones is significantly different compared to the older `Records.json` format. Here’s a comparison for clarity:

**Old Format (`Records.json`):**
```json
{
  "locations": [
    {
      "latitudeE7": 516553765,
      "longitudeE7": 40632229,
      "accuracy": 24,
      "timestamp": "2012-01-27T21:14:42.352Z"
    }
  ]
}
```

**New Format (Exported from Phone):**
```json
{
  "rawSignals": [
    {
      "position": {
        "LatLng": "51.6553765°, 4.0632229°",
        "accuracyMeters": 24,
        "altitudeMeters": 65.3,
        "source": "UNKNOWN",
        "timestamp": "2024-12-07T17:46:25.000+01:00",
        "speedMetersPerSecond": 0.0
      }
    }
  ]
}
```

## Observations
The structure of the new file is not compatible with the previous implementation of `gotoextr`, which expects the older locations array.
The panic error (invalid memory address or nil pointer dereference) is unexpected and should not occur, regardless of the input file format.

## Next Steps
I will address the panic crash to ensure `gotoextr` handles unexpected file structures gracefully.
I will also explore making `gotoextr` compatible with the new JSON format, so it can correctly parse and convert the newer location data.

## Workaround for Now
Unfortunately, due to the incompatible format, there is no immediate solution to convert your data. However, I’ll prioritize fixing this and notify you once the updates are available.

Thank you for your patience and for bringing this to my attention.

**Comment by wunderbaum** [comment]
Hello,
my whole location-takeout is about 6GB large. Maybe that is way to much. But when there is no records.json, thats not useful.
I took a deeper look in this Timeline Edits.json. I think its just that what it says. Every time where maps was unsure what I did, it asked my but I have not edited every single false entry.
```
"deviceId": "757593298",
    "rawSignal": {
      "signal": {
        "activityRecord": {
          "detectedActivities": [{
            "activityType": "UNKNOWN",
            "probability": 0.41
          }, {
            "activityType": "IN_VEHICLE",
            "probability": 0.1
          }, {
            "activityType": "ON_BICYCLE",
            "probability": 0.1
          }, {
            "activityType": "ON_FOOT",
            "probability": 0.1
          }, {
            "activityType": "STILL",
            "probability": 0.1
          }, {
            "activityType": "WALKING",
            "probability": 0.1
          }, {
            "activityType": "RUNNING",
            "probability": 0.1
          }, {
            "activityType": "IN_ROAD_VEHICLE",
            "probability": 0.1
          }, {
            "activityType": "IN_RAIL_VEHICLE",
            "probability": 0.1
          }],
          "timestamp": "2024-03-31T11:12:25.499Z"
        }
````

How have you exported the device stored data? I had only the option for an encrypted backup.

**Comment by kpym** [comment]
@wunderbaum could you tell me:
1. Does one of the json files inside the zip contains a 'locations' or 'rawSignals' array?
2. If so, what is the full name of that .json file?

**Comment by kpym** [comment]
To export the location from your phone you can do (following the instructions of perplexity) : 
`Settings -> Location -> Location Services -> Time Line -> Export Timeline Data`.

**Comment by wunderbaum** [comment]
> @wunderbaum could you tell me:
> 
> 1. Does one of the json files inside the zip contains a 'locations' or 'rawSignals' array?
> 2. If so, what is the full name of that .json file?

Unfortunately, there is no such file in that directory. I have a huge collection of all the pictures from menus and food I did :-)
In the Timeline-folder I have
```
Tomstones.csv
Timeline Edits.json
Settings.json
Encrypted Backups.txt
```

I managed to export my Zeitachse.json (Timeline.json). In that file, I can find rawSignals, but that seems to start in 2024. Earlier, the format was different:

```
"startTime": "2018-07-18T14:00:00.000+02:00",
      "endTime": "2018-07-18T16:00:00.000+02:00",
      "timelinePath": [
        {
          "point": "41.1829322°, 19.479344°",
          "time": "2018-07-18T14:38:00.000+02:00"
        },
        {
          "point": "41.1840254°, 19.481436°",
          "time": "2018-07-18T14:40:00.000+02:00"
        },
```

Interesting how precise I was tracked or tracked myself whatsoever. Seems that they are running out of storage tracking every android-user day by day :-)

**Comment by kpym** [comment]
I also have "semanticSegments" in my .json exported from my phone, but I also have "rawSignals" part. You don't have it?

And in your 6Gb .zip, what is the largest .json file? Can you check what are the top fields of that json?

**Comment by wunderbaum** [comment]
> I also have "semanticSegments" in my .json exported from my phone, but I also have "rawSignals" part. You don't have it?
> 
> And in your 6Gb .zip, what is the largest .json file? Can you check what are the top fields of that json?

The exported from the phone starts with _semanticSegments_ in the beginning. 
```
{
  "semanticSegments": [
    {
      "startTime": "2014-04-14T14:00:00.000+02:00",
      "endTime": "2014-04-14T16:00:00.000+02:00",
      "timelinePath": [
        {
          "point": 
```
In my exported one, the largest is that _Timeline Edits.json_

**Comment by kpym** [comment]
> > I also have "semanticSegments" in my .json exported from my phone, but I also have "rawSignals" part. You don't have it?
> > And in your 6Gb .zip, what is the largest .json file? Can you check what are the top fields of that json?
> 
> The exported from the phone starts with _semanticSegments_ in the beginning.

That was not the question. Mine also starts with "semanticSegments", but after that I have "rawSignals". Do you also have "rawSignals"?

> In my exported one, the largest is that _Timeline Edits.json_

This is some sort of log file of your position edits, I think. I don't know how to retrieve your position history from it.
Do you have other json files in the zip?

**Comment by wunderbaum** [comment]
No, I think the exported one from the phone is the only one I have.

**Comment by kpym** [comment]
I came to release a new version [v0.8.0](https://github.com/kpym/gotoextr/releases/tag/v0.8.0) that will hopefully be compatible with the new json format exported from the phone. 

Could you check it, please?

**Comment by wunderbaum** [comment]
P:\>gotoextr.exe -s 2014-01-01 -f kml  Zeitachse.json
Read 6064 positions in 0.48 seconds
Wrote 0 positions in 1 segments in 1 tracks

Reading seems to work, but the result is empty.
Let me take a nap, its 23h in my timezone.

**Comment by kpym** [comment]
This is the expected behaviour from `gotoextr`. The reason your exported location history from your Android phone doesn't contain points from 2014 is due to Google's significant change in how location data is stored, implemented in 2024. You can find more details about this change by consulting [Google's official sources](https://www.google.com/search?q=site%3Agoogle.com+Takout+Location+History). Based on [1][1], [2][2] and some other sites :

1. *Limited historical data:* Google moved Location History (now called "Timeline") from cloud storage to on-device storage. When this transition occurred, only recent data was transferred to the device. Older data, such as points from 2014, remained in the cloud but is no longer accessible through the usual export methods.
1. *Default data retention:* By default, the new on-device storage is set to retain only 3 months of data. This means that even more recent data might be automatically deleted unless you change the settings.
1. *Loss of long-term history:* As a result of this transition, users lost access to their long-term location history that was previously stored in the cloud.
 
To preserve older data, Google notified users via email before the change, giving an opportunity to save their historical data. Unfortunately, if you missed this step, recovering data from 2014 may no longer be possible.

To export all positions from your `Zeitachse.json`, you can use the following command:
```sh
> gotoextr -s 2001-01-01 -e 2025-01-01 -f kml Zeitachse.json
```

However, keep in mind that the exported data will likely contain only points from the last three months, as explained above.

**Notes:** 
1. When using `-s` without specifying `-e`, only one day is exported because the default value of `-e` is not `<today>` but matches the value of `<start>`.
2. Probably your `Zeitachse.json` contains less then 6064 valid points. There was a bug in `v0.8.0`, so I released [v0.8.1](https://github.com/kpym/gotoextr/releases/tag/v0.8.1) to fix this counting problem.

[1]: https://support.google.com/accounts/answer/3118687
[2]: https://blog.google/products/maps/updates-to-location-history-and-new-controls-coming-soon-to-maps/

**Comment by wunderbaum** [comment]
Thank you so much, I was able to convert:
```
F:\>gotoextr -s 2001-01-01 -e 2025-01-01 -f kml Zeitachse.json
Read 6064 positions in 0.49 seconds
Wrote 768 positions in 142 segments in 31 tracks
```

I uploaded the kml to earth.google.com but it seems that longitude and latitude are exchanged as I was never in the sea near african shore:

![afr](https://github.com/user-attachments/assets/4c78dab7-af82-44b2-bf7d-03ed953171d1)

Update: When I copy 2 single coordinates from the history in google, I get the right place. But the online kml-viewers mix it.

**Comment by kpym** [comment]
Thank you for pointing this out 🙏 

This is indeed a bug—actually a long-standing one. It seems that no one reported it before, likely because the `gotoextr` feature hasn't been used to export to KML. The issue arises from the coordinate format specified in KML, which uses `<Point><coordinates>lon,lat</coordinates></Point>`. This `lon,lat` order differs from the more human-readable `lat,lon` convention typically used in tools like Google Maps.

The bug has been addressed in the latest version, [v0.8.3](https://github.com/kpym/gotoextr/releases/tag/v0.8.3), where the coordinate order is now correctly handled.

Please update to the latest version and let me know if you encounter any further issues.

**Comment by wunderbaum** [comment]
Sorry, I was busy last few days.
Great! Now it works and I am in my area :-)
